internal/xds/rbac/matchers.go

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package rbac
+
+import (
+	"fmt"
+	"net"
+	"regexp"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3route_componentspb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	matcher "google.golang.org/grpc/internal/xds/matcher"
+)
+
+// TODO: Do we need errors at every node? Should we have checks on each node (i.e. destinationPort should be within a certain range)
+
+// The rbacMatcher interface will be used by the RBAC Engine to help determine
+// whether incoming RPC requests should be allowed or denied. There will be many
+// types of matchers, each instantiated with part of the policy used to
+// instantiate the RBAC Engine. These constructed matchers will form a logical
+// tree of matchers, which data about any incoming RPC's will be passed through
+// the tree to help make a decision about whether the RPC should be allowed or
+// not.
+type rbacMatcher interface {
```

Comment:
This is not ideal. Now, users of this type will have to use it as `rbac.rbacMatcher`. I would prefer naming this just `matcher`.

And the import renaming is caused because you have code like this:
```golang
	for _, matcher := range om.matchers {
		if matcher.matches(args) {
			return true
		}
	}
```

Here the local variable `matcher` collides with the package name `matcher`. So, if you change this code to
```golang
	for _, m := range om.matchers {
		if m.matches(args) {
			return true
		}
	}
```
you should be fine

Reply:
I don't understand. As long as you don't have a variable or field with the name `matcher`, you are not going to end up with any code which does `matcher.Foo` or `matcher.Foo()`. Why can't you avoid having such variables names or field names?

Reply:
I import the matcher package that I moved to internal/xds, so that is named matcher per your other comment. Now I have a proto import called matcher AND an interface called matcher.

Reply:
Talked about this in chat. Decided to rename import to "internalmatcher".